src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/CertificateAuthorityService.java

```diff
@@ -141,36 +141,48 @@ public void evict() {
      *                               false = only include top CAs
      * @throws InconsistentCaDataException if required CA data could not be extracted, for example due to OCSP
      * responses not being valid
-     * @return
+     * @return list of approved CAs
      */
     @Cacheable(GET_CERTIFICATE_AUTHORITIES_CACHE)
     public List<ApprovedCaDto> getCertificateAuthorities(KeyUsageInfo keyUsageInfo,
             boolean includeIntermediateCas) throws InconsistentCaDataException {
 
         log.info("getCertificateAuthorities");
         List<X509Certificate> caCerts = new ArrayList<>(globalConfService.getAllCaCertsForThisInstance());
+        caCerts.forEach(cert -> log.debug(String.format("Cert SubjectDN=%s IssuerDN=%s", cert.getSubjectDN(),
+                cert.getIssuerDN())));
+
         List<ApprovedCaDto> dtos = new ArrayList<>();
         // map of each subject - issuer DN pair for easy lookups
         Map<String, String> subjectsToIssuers = caCerts.stream().collect(
                 Collectors.toMap(
                         x509 -> x509.getSubjectDN().getName(),
                         x509 -> x509.getIssuerDN().getName()));
 
+        // we only fetch ocsp responses for intermediate approved CAs
+        // configured as approved CA and its issuer cert is also an approved CA
+        List<X509Certificate> filteredCerts = caCerts.stream()
+                .filter(cert -> subjectsToIssuers.containsKey(cert.getIssuerDN().getName()))
+                .collect(Collectors.toList());
+
         String[] base64EncodedOcspResponses;
         try {
-            String[] certHashes = CertUtils.getCertHashes(new ArrayList<>(caCerts));
+            String[] certHashes = CertUtils.getCertHashes(new ArrayList<>(filteredCerts));
             base64EncodedOcspResponses = signerProxyFacade.getOcspResponses(certHashes);
         } catch (Exception e) {
             throw new InconsistentCaDataException("failed to get read CA OCSP responses", e);
         }
-        if (caCerts.size() != base64EncodedOcspResponses.length) {
-            throw new InconsistentCaDataException("ocsp responses do not match ca certs");
+        if (filteredCerts.size() != base64EncodedOcspResponses.length) {
+            throw new InconsistentCaDataException(
+                    String.format("ocsp responses do not match ca certs %d vs %d",
+                            filteredCerts.size(), base64EncodedOcspResponses.length));
         }
 
         // build dtos
         for (int i = 0; i < caCerts.size(); i++) {
+            int idx = filteredCerts.indexOf(caCerts.get(i));
             dtos.add(buildCertificateAuthorityDto(caCerts.get(i),
-                    base64EncodedOcspResponses[i],
+                    (idx != -1) ? base64EncodedOcspResponses[idx] : null,
                     subjectsToIssuers));
         }
```

Comment:
Could be written as a "`for cert : caCerts`" loop, avoids the two get() calls and using the `i` by mistake.

Reply:
done